src/main.rs

```diff
@@ -44,19 +53,53 @@ async fn stylesheet() -> impl Responder {
         .body(STYLESHEET)
 }
 
+#[get("/{page}")]
+async fn page(path: web::Path<String>, data: web::Data<AppState>) -> impl Responder {
+    let path_name = path.into_inner();
+    info!("responding to GET at /{path_name}");
+    if let Some(content) = data.into_inner().pages.get(&path_name) {
+        Ok(templates::page(None, html! {(content)}))
+    } else {
+        Err(ErrorNotFound(""))
+    }
+}
+
 #[actix_web::main]
 async fn main() -> std::io::Result<()> {
     // Set up logging first
     tracing_subscriber::fmt()
         .with_max_level(tracing::Level::INFO)
         .init();
 
-    let server = HttpServer::new(|| {
+    // Read files from content directory
+    let mut pages = HashMap::new();
+    if let Ok(files) = read_dir(Path::new("content/")) {
+        for file in files {
+            let file_path = file.expect("IO error").path();
+            pages.insert(
+                file_path
+                    .file_stem()
+                    .expect("Path error")
+                    .to_str()
+                    .expect("String error")
+                    .to_string(),
+                read_to_string(file_path).expect("Error reading file contents"),
+            );
+        }
+    } else {
+        warn!("No pages found, content folder is missing or empty")
+    }
```

Comment:
Rather than guessing at what caused the error, a `match` here would let you log the error:

```suggestion
    match read_dir(Path::new("content/")) {
        Ok(files) => {
            for file in files {
                let file_path = file.expect("IO error").path();
                pages.insert(
                    file_path
                        .file_stem()
                        .expect("Path error")
                        .to_str()
                        .expect("String error")
                        .to_string(),
                    read_to_string(file_path).expect("Error reading file contents"),
                );
            }
        }
        Err(err) => {
            warn!(%err, "failed to read pages from content folder")
        }
    }
```

Reply:
To go a little bit further, though: I would recommend extracting this logic out into a separate function, whose signature would look something like this...

```rust
fn read_content(path: &Path) -> Result<HashMap<String, String>, std::io::Error> { }
```

...and which you could call like this...

```rust
let pages = match read_content(&Path::new("content/")) {
    Ok(files) => files,
    Err(err) => {
        warn!(%err, "failed to read pages from content folder")
    }
}
```

My reason for suggesting this is that, because the `read_content()` function returns a `Result`, you should be able to replace all the calls to `.expect()` inside its body with usages of `?`, which will early-return the error. Then, your `match` on the call to `read_content()` serves as a single-point handling and logging of the error, which is more uniform and will not crash the app.